### PR TITLE
LUCENE-8282: Reduce boxing and unnecessary object creation in DV updates

### PR DESCRIPTION
**Pull request by s1monw**
DV updates used the boxed type Long to keep API generic. Yet, the missing
type caused a lot of code duplication, boxing and unnecessary object creation.
This change cuts over to type safe APIs using BytesRef and long (the primitive)

In this change most of the code that is almost identical between binary and numeric
is not shared reducing the maintenance overhead and likelihood of introducing bugs.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/index/FrozenBufferedUpdates.java [hunk 1]
```diff
@@ -641,48 +595,39 @@ private long applyDocValuesUpdates(BufferedUpdatesStream.SegmentState segState,
...
-            }
-          }
+          IntPredicate limitPredicate = doc -> segState.rld.sortMap.newToOld(doc) < limit;
+          addPredicate = isLivePredicate == null ? limitPredicate : isLivePredicate.and(limitPredicate);
         } else {
-          int doc;
-          final Bits acceptDocs = segState.rld.getLiveDocs();
-          while ((doc = postingsEnum.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
-            if (doc >= limit) {
-              break; // no more docs that can be updated for this term
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/index/FrozenBufferedUpdates.java]: could this break have been important for performance of updates?

Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/index/FrozenBufferedUpdates.java]: yeah true! I think I will roll back that part

**Comment by s1monw** [comment]
> I suspect object creation didn't hurt too much with espace analysis since it looks like the boxed longs never escaped, but I like the better type safety.

I am not sure that is the case. We used a boxed value that is held in maps etc. that surely escaped here. Some uses of it might have been ok but not all of them.

**Comment by s1monw** [comment]
@shaie thanks I had some unpushed changes can you take another look?

**Pull request merged**